### PR TITLE
Scripts to rebrand variables.env

### PR DESCRIPTION
**Pull request by mserranom**
## Description

Adds a `bin/update-env` script to rebrand environment variables from `SHARELATEX_` to `OVERLEAF_`, and updates `bin/up` and `bin/upgrade` with warnings and prompts to upgrade.

Variables in `lib/config-seed/variables.env` should be updated along with `lib/config-seed/version` once CE/SP v5 are published.

List of changes:

**`bin/update-env`**

Updates `SHARELATEX_` prefix with `OVERLEAF_` in `/config/variables.env`.

```
$ cat config/variables.env | grep 'SHARELATEX_' 
71

$ bin/update-env 
This script will update your config/variables.env. 
We recommend backing up your config with bin/backup-config.
Do you want to continue? (y/n): n
Exiting.

$ bin/update-env
This script will update your config/variables.env.
We recommend backing up your config with bin/backup-config.
Do you want to continue? (y/n): y
Creating backup file config/__old-variables.env
Replacing 'SHARELATEX_' with 'OVERLEAF_' in config/variables.env
Done.

$ cat config/variables.env | grep 'SHARELATEX_' 
0
```

**`bin/up`**

Prompts a warning when major version `>=5` and there are environment variables containing `SHARELATEX_` in `config/variables.env`

```
$ echo 5.0.0 > config/version
$ bin/up
WARNING: some environment variables defined in config/variables.env
contain 'SHARELATEX_' in their name
Starting with Overleaf CE and Server Pro 5.0.0 the environment variables
use the prefix 'OVERLEAF_'  instead of '_SHARELATEX'.
Please check your config/version and config/variables.env files
Do you want to continue? (y/n): n
Exiting.

$ bin/up                     
WARNING: some environment variables defined in config/variables.env
contain 'SHARELATEX_' in their name
Starting with Overleaf CE and Server Pro 5.0.0 the environment variables
use the prefix 'OVERLEAF_'  instead of '_SHARELATEX'.
Please check your config/version and config/variables.env files
Do you want to continue? (y/n): y
Initiating Mongo replica set...

$ echo 4.2.0 > config/version
$bin/up                     
Initiating Mongo replica set...            
```

**`bin/upgrade`**

When upgrading to `5.x` from a lower major version prompts the user to rename the environment variables in `config/variables.env`.

Tested by faking `5.0.0` in `lib/config-seed/version`

```
$ cat config/variables.env | grep 'SHARELATEX_' | wc -l
71

$ echo 5.0.0 > lib/config-seed/version      
$ echo 4.2.0 > config/version                          
           
$ bin/upgrade                                          
Warning: current branch is not master, 'msm-env-update-script' instead
Checking for code update...
No code update available for download
New docker image version available (5.0.0)
Current image version is '4.2.0' (from config/version)
WARNING: this is a major version update, please check the Release Notes for breaking changes before proceeding:
* https://github.com/overleaf/overleaf/wiki#release-notes
Upgrade image? [y/n] y
Upgrading config/version from 4.2.0 to 5.0.0
At this point, we recommend backing up your data before proceeding
!! WARNING: Only do this while the docker services are stopped!!
Proceed with the upgrade? [y/n] y
Backing up old version file to config/__old-version
Over-writing config/version with 5.0.0
Renaming environment variables in config/variables.env
Proceed with the environment variable renaming? (y/n): y
Backing up old variables.env file to config/__old-variables.env
Done

$ cat config/variables.env | grep 'SHARELATEX_' | wc -l
0
$ cat config/__old-variables.env | grep 'SHARELATEX_' | wc -l 
71
```

## Contributor Agreement

- [x] I confirm I have signed the [Contributor License Agreement](https://github.com/overleaf/overleaf/blob/master/CONTRIBUTING.md#contributor-license-agreement)

### CODE REVIEW

**Comment by mserranom** [comment]
@das7pad @briangough let me know if you think there should be extra checks or migration paths.

File: bin/update-env [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+source "$TOOLKIT_ROOT/lib/shared-functions.sh"
+
+function usage() {
+  echo "Usage: bin/update-env"
+  echo ""
+  echo "Updates config/variables.env, adding new environment variables"
+  echo ""
+  echo "This program will pass any extra flags to docker compose,"
+  echo "for example: 'bin/up -d' will run in detached mode"
```
Comment (das7pad) [inline bin/update-env:27]:
```suggestion
```

File: bin/update-env [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    exit
+  fi
+
+  # read_image_version
+  echo "This script will update your config/variables.env."
+  echo "We recommend backing up your config with bin/backup-config."
+  prompt "Do you want to continue?"
+
+  echo "Creating backup file config/__old-variables.env"
+  cp config/variables.env config/__old-variables.env
```
Comment (das7pad) [inline bin/update-env:42]: If I run this twice, my earlier backup would be gone. WDYT about adding a timestamp into the path?

**Review by das7pad — COMMENTED** [review]
As for the name of the new script, WDYT about being more explicit and calling it `bin/rename-env-vars-5-0`?

File: bin/upgrade [hunk 3]
```diff
@@ -148,9 +155,17 @@ function handle_image_upgrade() {
...
-  echo "Over-writing config/version with $seed_image_version"
+  echo "Over-writing config/version with $SEED_IMAGE_VERSION"
   cp "$TOOLKIT_ROOT/lib/config-seed/version" "$TOOLKIT_ROOT/config/version"
 
+  if [[ "$IMAGE_VERSION_MAJOR" -le 4 && "$SEED_IMAGE_VERSION_MAJOR" -ge 5 ]]; then 
+    echo "Renaming environment variables in config/variables.env"
+    prompt "Proceed with the environment variable renaming?"
+    echo "Backing up old variables.env file to config/__old-variables.env"
+    cp "$TOOLKIT_ROOT/config/variables.env" "$TOOLKIT_ROOT/config/__old-variables.env"
+    sed -i "s/SHARELATEX_/OVERLEAF_/g" "$TOOLKIT_ROOT/config/variables.env"
```
[the remark this answers is not in the document]
Comment (mserranom) [reply bin/upgrade:166]: Yeah, they're pretty much the same. If there are no other suggestions for these lines and the migration script I'll do the change.

File: bin/update-env (hunk 2 above)
Comment (mserranom) [reply bin/update-env]: Done in https://github.com/overleaf/toolkit/pull/214/commits/8b9fc04f55ff5d93a9c0e0732a9a1200a091e967

File: bin/upgrade (hunk 3 above)
[the remark this answers is not in the document]
Comment (mserranom) [reply bin/upgrade]: Ended up not doing it, prompts are different, and doing the plumbing to offer different types of prompts feels overkill just for 2 commands.

**Comment by mserranom** [comment]
> LGTM, is the script name still on your radar (see my last review summary)?

Yes, this is now done (used your suggestion).

**Review by briangough — COMMENTED** [review]
For maintainability, how about moving the file backup logic into shared functions?  

The sed command could also go in there.  

Some suggestions for the sed command:

1. check if there will be any changes needed at the start and bail out with a "No changes needed" message otherwise (no backup needed in that case).  This check could ignore commented lines, to avoid unnecessary runs of the script.
2. show the lines to be changed with `sed -n '/SHARELATEX_/p' `  before running the update?
3. report the number of lines changed when the update finishes

**Comment by das7pad** [comment]
> This check could ignore commented lines, to avoid unnecessary runs of the script.

I suggest migrating these as well, otherwise things might not work as expected after restoring the line.

**Comment by briangough** [comment]
> I suggest migrating these as well, otherwise things might not work as expected after restoring the line.

Right, I see from looking at `lib/config-seed/variables.env` that our default is to have all the `SHARELATEX_` definitions commented out so I agree with modifying the comments in this case.

**Comment by mserranom** [comment]
> For maintainability, how about moving the file backup logic into shared functions?
> 
> The sed command could also go in there.
> 
> Some suggestions for the sed command:
> 
> 1. check if there will be any changes needed at the start and bail out with a "No changes needed" message otherwise (no backup needed in that case).  This check could ignore commented lines, to avoid unnecessary runs of the script.
> 2. show the lines to be changed with `sed -n '/SHARELATEX_/p' `  before running the update?
> 3. report the number of lines changed when the update finishes

Done in https://github.com/overleaf/toolkit/pull/214/commits/9c5ac3a568775842103845cf622effe41b427fbd.

**Review by briangough — APPROVED** [review]

**Pull request merged**